### PR TITLE
FW 2529 Add the preview to the links shared via direct link [Chat widget]

### PR DESCRIPTION
**Pull request by Ojashwa**
### What do you want to achieve?
- Add the preview to the shared links in chat widget 

### PR Checklist

- [x] I have tested it locally and all functionalities are working fine.
- [x] I have compared it with mocks and all design elements are the same.
- [x] I have tested it in IE Browser.

### How was the code tested?

---

https://user-images.githubusercontent.com/22856546/127298400-a2b2770a-95f3-4a64-b1e2-ea995c76c2cb.mov

### What new thing you came across while writing this code? 
-

### In case you fixed a bug then please describe the root cause of it? 
-

NOTE: Make sure you're comparing your branch with the correct base branch

### CODE REVIEW

**Review by shubhamtibra — DISMISSED** [review]

**Review by devashishmamgain — CHANGES_REQUESTED** [review]
looks stretched

**Comment by devashishmamgain** [comment]
@Ojashwa @sritichamola pls review it, width and height ratio seems to be a problem

Screenshot 2021-07-28 at 5 45 07 PM

**Comment by Ojashwa** [comment]
> 
> 
> @Ojashwa @sritichamola pls review it, width and height ratio seems to be a problem
> 
> Screenshot 2021-07-28 at 5 45 07 PM

- @devashishmamgain Please verify the image size.
---
image
---
image

**Comment by devashishmamgain** [comment]
> > @Ojashwa @sritichamola pls review it, width and height ratio seems to be a problem
> > Screenshot 2021-07-28 at 5 45 07 PM
> 
> * @devashishmamgain Please verify the image size.
> 
> ## image
> image

is it deployed anywhere? want to test it
@Ojashwa

File: server/routers/routes.js [hunk 1]
```diff
@@ -22,4 +22,18 @@ home.get('/kommunicate.app', function (req, res) {
...
 home.get('/:version/kommunicate.app', webpluginController.getPlugin);
 
-home.get('/chat', webpluginController.getPluginHTML)
\ No newline at end of file
+home.get('/chat', webpluginController.getPluginHTML)
+home.get('/extractlink',function(req, res){
+    var urlToExtract = req.query.linkToExtract; 
+    if(!req.query.url){
+        res.status(400);
+    }
```
Comment (Prochnost55) [inline server/routers/routes.js:30]: Where is req.query.url used?

File: webplugin/js/app/km-richtext-markup-1.0.js [hunk 2]
```diff
@@ -200,7 +200,20 @@ Kommunicate.markup = {
             </div>`
         );
     },
-
+    getLinkPreviewTemplate: function (extractedData) {
+        var data = extractedData.data;
+        return (
```
Comment (Prochnost55) [inline webplugin/js/app/km-richtext-markup-1.0.js:205]: can data be empty?

File: webplugin/js/app/mck-sidebox-1.0.js [hunk 3]
```diff
@@ -9171,29 +9171,41 @@ var userOverride = {
...
-                            });
+                    KommunicateUI.getLinkDataToPreview(
+                        msg.message,
+                        function (template) {
+                            $textMessage.prepend(template);
                         }
-                        $textMessage.append(x);
+                    );
+                } 
+                // else {
```
Comment (Prochnost55) [inline webplugin/js/app/mck-sidebox-1.0.js:9184]: we can remove comments

File: webplugin/js/app/km-richtext-markup-1.0.js (hunk 2 above)
Comment (Ojashwa) [reply webplugin/js/app/km-richtext-markup-1.0.js:205]: it will return an object containing metadata of URL.

**Review by Prochnost55 — APPROVED** [review]

File: package.json [hunk 4]
```diff
@@ -28,6 +28,7 @@
     "ejs": "^3.1.6",
     "express": "4.17.1",
     "form-data": "3.0.0",
+    "link-preview-js": "^2.1.8",
```
Comment (devashishmamgain) [inline package.json:31]: remove ^ so that it doesn't use the latest version which might go without the testing

File: webplugin/js/app/km-richtext-markup-1.0.js [hunk 5]
```diff
@@ -200,7 +200,22 @@ Kommunicate.markup = {
...
+            return (
+                '<div class="link-preview-wrapper"><div><img class="link-preview-image" src="' +
+                (data.images[0] || data.favicons[0]) +
+                '" alt="' +
+                (data.siteName || data.title) +
+                '" width="80" height="80"></div><div class="link-preview-content"><h5 class="link-preview-title"> ' +
+                (data.siteName || data.title) +
+                '</h5><p class="link-preview-div-description">' +
+                (data.description || data.title) +
+                '</p></div></div>Here is the link:'
```
Comment (devashishmamgain) [inline webplugin/js/app/km-richtext-markup-1.0.js:215]: Here is the link: 
is same for all languages? this need to be in MCK_LABELS?

File: webplugin/js/app/kommunicateCommons.js [hunk 6]
```diff
@@ -108,7 +108,14 @@ function KommunicateCommons() {
         if (!object) return false;
         return typeof object == 'object' && object.constructor == Object;
     };
-    
+    _this.isUrlValid = function (url) {
+        if (!url) return false;
+        return /^(?:(?:(?:https?|ftp):)?\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,})))(?::\d{2,5})?(?:[/?#]\S*)?$/i.test(
+            url
+        )
+            ? true
```
Comment (devashishmamgain) [inline webplugin/js/app/kommunicateCommons.js:116]: ? true: false part is not required.

**Review by devashishmamgain — CHANGES_REQUESTED** [review]

**Review by devashishmamgain — CHANGES_REQUESTED** [review]
is it the server side plugin? where API call will come to the server?
can you find out a way to do it on client side itself, problem with server side is that then server side scaling will be required and also in future we wouldn't be able to make the chat widget plugin as pure javascript based.

currently it is some API call .app related but we would want to cut down that dependency too and make it light weight with just javascript file, that can be served from CDN also in future.

@gargshantnu can you review this once

**Comment by gargshantnu** [comment]
> is it the server side plugin? where API call will come to the server?
> can you find out a way to do it on client side itself, problem with server side is that then server side scaling will be required and also in future we wouldn't be able to make the chat widget plugin as pure javascript based.
> 
> currently it is some API call .app related but we would want to cut down that dependency too and make it light weight with just javascript file, that can be served from CDN also in future.
> 
> @gargshantnu can you review this once

right @devashishmamgain. having an API in this code will create bottleneck and performance issues.
also, as library mentioned:

> **This library fetches the website and parses its html, as if the user would visit the page**

If page is heavy, it might create more problems on server.

@Ojashwa please look for some other approach.

**Comment by Ojashwa** [comment]
> 
> 
> > is it the server side plugin? where API call will come to the server?
> > can you find out a way to do it on client side itself, problem with server side is that then server side scaling will be required and also in future we wouldn't be able to make the chat widget plugin as pure javascript based.
> > currently it is some API call .app related but we would want to cut down that dependency too and make it light weight with just javascript file, that can be served from CDN also in future.
> > @gargshantnu can you review this once
> 
> right @devashishmamgain. having an API in this code will create bottleneck and performance issues.
> also, as library mentioned:
> 
> > **This library fetches the website and parses its html, as if the user would visit the page**
> 
> If page is heavy, it might create more problems on server.
> 
> @Ojashwa please look for some other approach.

- Did some research to impliment it on the client end and found one solution with Facebook Graph API. But the data we get from Facebook is some time is broken or not upto thr mark will attach some refrances below:
---
image
---
image
---
- In below screen shot image URl is missing in response.
image
---
image

Also, Facebook Graph API will follow The Open Graph [protocol ](https://ogp.me/)and share data of URL if that URL consist basic metadata. We might not get data for every URL.
cc: @devashishmamgain @gargshantnu

**Comment by devashishmamgain** [comment]
> > > is it the server side plugin? where API call will come to the server?
> > > can you find out a way to do it on client side itself, problem with server side is that then server side scaling will be required and also in future we wouldn't be able to make the chat widget plugin as pure javascript based.
> > > currently it is some API call .app related but we would want to cut down that dependency too and make it light weight with just javascript file, that can be served from CDN also in future.
> > > @gargshantnu can you review this once
> > 
> > 
> > right @devashishmamgain. having an API in this code will create bottleneck and performance issues.
> > also, as library mentioned:
> > > **This library fetches the website and parses its html, as if the user would visit the page**
> > 
> > 
> > If page is heavy, it might create more problems on server.
> > @Ojashwa please look for some other approach.
> 
> * Did some research to impliment it on the client end and found one solution with Facebook Graph API. But the data we get from Facebook is some time is broken or not upto thr mark will attach some refrances below:
> 
> ## image
> ## image
> * In below screen shot image URl is missing in response.
>   image
> 
> image
> 
> Also, Facebook Graph API will follow The Open Graph [protocol ](https://ogp.me/)and share data of URL if that URL consist basic metadata. We might not get data for every URL.
> cc: @devashishmamgain @gargshantnu

search for more javascript plugins

File: server/config/config-env.js [hunk 7]
```diff
@@ -18,7 +18,7 @@ const configEnv = {
             },
             aws: {
                 cdnUrl: 'https://cdn-test.kommunicate.io',
-                bucket: 'km-test-cdn',
+                bucket: 'cdn-kommunicate',
```
Comment (devashishmamgain) [inline server/config/config-env.js:21]: is this change related to this issue?

Comment (priyadarshi) [reply server/config/config-env.js:21]: It needs to revert as we have migrated test cdn to new aws account

Comment (priyadarshi) [reply server/config/config-env.js:21]: @Ojashwa Please use bucket: 'km-test-cdn'

Comment (Ojashwa) [reply server/config/config-env.js:21]: > 
> 
> is this change related to this issue?

@priyadarshi sir could you please describe the changes?

Comment (priyadarshi) [reply server/config/config-env.js:21]: yes

**Review by devashishmamgain — APPROVED** [review]

**Review by shubhamtibra — APPROVED** [review]

**Review by Prochnost55 — APPROVED** [review]
